Add doc for URL binding feature (#1346)
symfony/ux: src/LiveComponent/doc/index.rst

    use Symfony\UX\LiveComponent\DefaultActionTrait;

    #[AsLiveComponent]
    class SearchModule
    {
        use DefaultActionTrait;

        #[LiveProp(writable: true, url: true)]
        public string $query = '';
    }

Now if you change the value of the ``query`` prop, the url will be updated to reflect the new state of your
component, for example: ``https://my.domain/search?query=my+search+string``.

Then, if you load this URL in your browser, your component will be initialized with the values provided in the query
string.

Supported data types
~~~~~~~~~~~~~~~~~~~~

You can use scalars, arrays and objects in your URL bindings:

============================================  =================================================
``prop`` value                                URL representation
============================================  =================================================
``'some search string'``                      ``prop=some+search+string``
``42``                                        ``prop=42``
``['foo', 'bar']``                            ``prop[0]=foo&prop[1]=bar``
``{ foo: 'bar', baz: 42 }``                   ``prop[foo]=bar&prop[baz]=42``

.. note::

>     Type consistency is enforced by the LiveComponent internal hydrator when the component is initialized with query

Reviewer: > Type consistency is enforced by the LiveComponent internal hydrator. When the component is initialized, If a query parameter could not be hydrated properly into a LiveProp, it will be ignored to avoid unfriendly errors for the end user.

Just a suggestion.. feel free to ignore if you prefer the original ;)

Author: I prefer the original, I find it more accurate as this type consistency check is only used on component initialization. But thanks for the suggestion :)

Reviewer 2: I think we should make this not in a note, rather a normal sentence, as it's really important. And, from a user's point of view, the important thing is that it goes through the hydration system. Something like:

> When a page is loaded with a query parameter that's bound to a ``LiveProp`` (e.g. ``/search?query=my+search+string``), the value - ``my search string`` - goes through the hydration system before it's set onto the property. If a value can't be hydrated, it will be ignored.

    parameters. If a value could not be hydrated properly, it will be ignored to avoid unfriendly errors for the end
    user.

Multiple bindings
~~~~~~~~~~~~~~~~~

You can use as many URL bindings as you want in your component. To ensure the state is fully represented in the URL,

> all bound props will be added to query parameters, even if their values didn't change.

Reviewer: "added" may be missleading there i think...    "set as" ?

For example, if you declare the following bindings::

    // ...
    #[AsLiveComponent]
    class SearchModule
    {
        #[LiveProp(writable: true, url: true)]
        public string $query = '';

        #[LiveProp(writable: true, url: true)]
        public string $mode = 'fulltext';

        // ...
    }

And you only set the ``query`` value, then your URL will be updated to ``https://my.domain/search?query=my+query+string&mode=fulltext``.

Validation
~~~~~~~~~~

> To validate your component state when it is rendered for the first time, you have to setup a `PostMount hook`_::

Reviewer: I'd use a warning or caution block there..  wdyt ?

Author: I added a caution block at the top of the section

Reviewer 2: `setup` -> `set up`

        #}
    >

> Binding LiveProp to URL query parameters

Reviewer 2: How about:

> Changing the URL when a LiveProp changes

----------------------------------------

.. versionadded:: 2.13

>     The ``url`` option was introduced in Live Components 2.13.

Reviewer 2: 2.14

> You can bind your component LiveProp's values to URL query parameters thanks to the ``url`` option::

Reviewer 2: How about:

> If you want the URL to update when a ``LiveProp`` changes, you can do that with the ``url`` option:

>     // src/Components/SearchModule.php

Reviewer 2: ```
    // src/Twig/Components/SearchModule.php
```

right?

Author: Other examples in the doc live in `src/Components`, not sure why this one should be different?

    namespace App\Components;

    use Symfony\UX\LiveComponent\Attribute\AsLiveComponent;
    use Symfony\UX\LiveComponent\Attribute\LiveProp;
    use Symfony\UX\LiveComponent\DefaultActionTrait;

    #[AsLiveComponent]
    class SearchModule
    {
        use DefaultActionTrait;

        #[LiveProp(writable: true, url: true)]
        public string $query = '';
    }

> Now if you change the value of the ``query`` prop, the url will be updated to reflect the new state of your

Reviewer 2: > Now, when the user changes the...

> ... prop, a query parameter in the URL will be updated...

component, for example: ``https://my.domain/search?query=my+search+string``.

> Then, if you load this URL in your browser, your component will be initialized with the values provided in the query

Reviewer 2: ```suggestion
If you load this URL in your browser, the ``LiveProp`` value will be initialized using the query string (e.g. ``my search string``).
```

string.

.. note::

>     URL is updated in place, no new entry is added to the browser's history.

Reviewer 2: The URL is changed via ``history.replaceState()``. So no new entry is added...

.. warning::

    You can use multiple components with URL bindings in the same page, as long as bound field names don't collide.

>     Otherwise, you will observe unexpected behaviors.

Reviewer 2: Someone may have suggested adding this, I say remove it. I think this is self-evident (e.g. if you have 2 variables defined in the same function with the same name, unexpected things will happen).

Author: Do you suggest removing the full warning block or just the last line?

> Supported data types

Reviewer 2: ```suggestion
Supported Data Types
```

~~~~~~~~~~~~~~~~~~~~

> You can use scalars, arrays and objects in your URL bindings:

Reviewer 2: You say objects but then don't show an example below.

Author: That was the purpose of the last example: `{ foo: 'bar', baz: 42 }`. Those values are the "serialized' ones, maybe I should change the table header to something like `JavaScript prop values`?

============================================  =================================================
``prop`` value                                URL representation
============================================  =================================================
``'some search string'``                      ``prop=some+search+string``
``42``                                        ``prop=42``
``['foo', 'bar']``                            ``prop[0]=foo&prop[1]=bar``
``{ foo: 'bar', baz: 42 }``                   ``prop[foo]=bar&prop[baz]=42``

.. note::

    Type consistency is enforced by the LiveComponent internal hydrator when the component is initialized with query
    parameters. If a value could not be hydrated properly, it will be ignored to avoid unfriendly errors for the end
    user.

> Multiple bindings

Reviewer 2: ```suggestion
Multiple Query Parameter Bindings
```

~~~~~~~~~~~~~~~~~

You can use as many URL bindings as you want in your component. To ensure the state is fully represented in the URL,
all bound props will be set as query parameters, even if their values didn't change.

For example, if you declare the following bindings::

    // ...
    #[AsLiveComponent]
    class SearchModule
    {
        #[LiveProp(writable: true, url: true)]
        public string $query = '';

        #[LiveProp(writable: true, url: true)]
        public string $mode = 'fulltext';

        // ...
    }

And you only set the ``query`` value, then your URL will be updated to ``https://my.domain/search?query=my+query+string&mode=fulltext``.

> Validation

Reviewer 2: ```suggestion
Validating the Query Parameter Values
```

~~~~~~~~~~

.. caution::

    It is recommended to validate data provided through query string parameters, so you can prevent malicious inputs

>     from being processed and rendered.

Reviewer 2: This is interesting. If a `LiveProp` is writable, then the fact that it is bound to a query parameter does not open a new malicious entrypoint: the value can always be changed to "anything". Two things:

A) If we want to talk about validating writable properties in general more clearly, then we can talk about that in another PR
B) This section may still make sense, but I think it's only important for the (probably rare) situation where you have `url: true` and `writable: false` (e.g. when the user calls a `LiveAction`, you change the prop inside that method, but the user doesn't change it directly).

Author: The goal of this section was to be explicit about the fact that bound props are not validated when initialized from the URL, even if the component uses validation. But yes, the security concern only applies when the validation logic itself involves some security process (sanitizing or whatever). Maybe I can change the caution block to a note like:

`If your component uses validation, it is recommended to also validate data provided through query string parameters.`

For the B) part, I could add another example in this section to illustrate this use case, wdyt?

Reviewer 2: Hmm, ok, yes, it's more clear to me. So the messaging would be something like:

> Like any writable `LiveProp`, because the user can modify this value, you should consider adding validation (<-- link to that section). When you bind a `LiveProp` to the URL, the initial value is *not* automatically validated. To validate it: (then the code).

Also, it's beyond the scope of the doc, but should the URL-binding feature validate automatically? And reject the value (skip it) if it's invalid?

To validate your component state when it is rendered for the first time, you have to setup a `PostMount hook`_::

    // ...
    use Symfony\Component\Validator\Constraints as Assert;
    use Symfony\UX\LiveComponent\ValidatableComponentTrait;
    use Symfony\UX\TwigComponent\Attribute\PostMount;

    #[AsLiveComponent]
    class SearchModule
    {
        use ValidatableComponentTrait;

        #[LiveProp(writable: true, url: true)]
        public string $query = '';

        #[LiveProp(writable: true, url: true)]
        #[Assert\NotBlank]
        public string $mode = 'fulltext';

        #[PostMount]
        public function postMount(): void
        {
            // Validate 'mode' field without throwing an exception

>             $this->validateField('mode', false);

Reviewer 2: It's not clear to me what happens when this fails.
